docs/structured.md

```diff
@@ -0,0 +1,212 @@
+  Columns = [{ ColumnID = …, Name = ... }, ...],
+  Indexes = [{ IndexID = …, Name = …, ColumnIDs = [ … ]}, ...],
+  Permissions = …
+  NextFreeColumnId = …
+  NextFreeIndexID = ...,
+  ...}
+```
+
+In order to support fast renaming, we indirect entity access through a
+hierarchical name-to-identifier mapping. NamespaceID and TableID are
+allocated from the same global ID space. ColumnID and IndexID are local
+to each Table.
+
+To simplify our implementation, our Tables will require the presence of a
+primary key. Follow-up work may relax this requirement. An investigation of use
+cases not requiring a primary key is required to specify this work.
+
+Initially, all the Table metadata in a cluster will be distributed by gossip and
+cached by each cockroach node. To reduce contention on the Table metadata, a
+database transaction on the data within a Table will only contain the data being
+addressed, and will not include a database read of the Table metadata. Care will
+be taken to not expose these implementation details to the user.
+
+##Addressing: Anatomy of a key##
+
+The “/” separators used below are shorthand for ordered encoding of the
+separated values.
+
+###Global metadata key###
+
+The root namespace is an unnamed namespace with a fixed ID of 0. Within it,
+metadata addressing will work as follows:
+
+Namespace: `\x00ns<NamespaceName>` -> NamespaceDescriptor { NamespaceID = …, ...}
+
+Table: `\x00tbl<NamespaceID><TableName>` -> TableDescriptor { TableID = …,
+local-IDs, … }
+
+Creating/Renaming/Deleting: Creating a new Namespace or Table involves
+allocating a new ID, initializing the {Namespace,Table}Descriptor, and storing
+the descriptor at key:/parent-id/name. Renaming involves deleting the
+`/old-parent-id/old-name` and creating `/new-parent-id/new-name`. A namespace can
+only be deleted if it does not contain any children. A table is deleted by
+marking it deleted in its table descriptor. This will allow folks to recover
+their data for a few days/weeks before we garbage collect the table in the
+background.
+
+###Data addressing###
+
+The Index data used to find data in the database is stored in the database
+itself at keys with the following prefix: `/TableID/IndexID/Key>`, where TableID
+represents the Table being addressed and IndexID the index in use.
+
+**Primary key addressing**
+
+This schema is made possible by the uniqueness constraint inherent to primary
+keys. The use of a primary key prefix: `/TableID/PrimaryIndexID/Key`, keys into a
+unique row in the database. A cell within a row under a particular column will
+be addressed by suffixing the desired ColumnID:
+`/TableID/PrimaryIndexID/Key/ColumnID`.  PrimaryIndexID is variable to allow
+changing the primary key.
+
+**Secondary key addressing**
+
+Secondary keys will be implemented as a layer of indirection to the primary
+keys. Unlike primary keys, secondary keys are not required to be unique.  So we
+will employ the following key anatomy that address null values:
```

Comment:
re: two comments up: It's even VERY useful immediately for things like indexing in lowercase so that the Microsoft example below can actually lookup "kimball" even if the entry says "Kimball".